Fix compilation with cereal >= 0.5

sdroege opened the pull request:
See https://github.com/acid-state/acid-state/pull/63 , which works around it by requiring older cereal versions.

sdroege commented on src/Data/Acid/Archive.hs:
```diff
@@ -18,10 +18,11 @@ import           Data.Acid.CRC
 
 import qualified Data.ByteString        as Strict
 import qualified Data.ByteString.Lazy   as Lazy
+import           Data.ByteString.Lazy.Builder
```
Btw, this would become Data.ByteString.Builder if you want to raise the bytestring dependency a bit. The Data.ByteString.Lazy.Builder is only provided for compatibility right now and deprecated

sdroege commented on src/Data/Acid/Archive.hs:
```diff
@@ -39,9 +40,10 @@ entriesToListNoFail Fail{}            = []
+    = execPut $ do
+        putWord64le contentLength
+        putWord16le contentHash
+        putLazyByteString content
```
Here it should also be possible to do the following without cereal at all. Let me know if you prefer that (and also the bytestring part above about the import). I'll update the pull request then.

    word64LE contentLength <>
    word16LE contentHash <>
    lazyByteString content

ddssff commented:
Does this patch work for cereal<0.5?  Does it need ifdefs?   Would the alternative version work with both?

sdroege commented:
The alternative version would, it just depends on bytestring's builder. This version should work at least with >= 0.4 (didn't check older docs) if I read the docs correctly, it's just a different Builder (the one from cereal instead of bytestring) that is going to be created there.

Your call which one you prefer, this version would have to be tested with cereal < 0.5 though, I can probably do that some time tomorrow.

stepcut commented:
cereal 0.5 also changed the way Float/Doubles are serialized. I have just now uploaded a new safecopy 0.9 which supports migration from the old format to the new format.

It would be mildly nice if acid-state could work with safecopy 0.8 or 0.9 (which depend on cereal 0.4 and cereal 0.5 respecitively) so that developers can have more flexibility on when they want to switch to the new Double/Float format.

Travis CI now tests against several versions of GHC. I believe the GHC 7.8.3 build will still use cereal 0.4.

I assume the new `putEntry` generates the same serialized data? I do not see any modifications to the `getEntry` function (or whatever it is called), so I assume we should not have any trouble reading old Entries?

It looks like the strict `Data.ByteString.Builder` was added in 0.10.2 which was uploaded in 2012. I have no objections to raising that lower bounds and using the non-deprecated builder.

sdroege commented:
I updated the pull request to also work with 0.4.1.0, it builds fine with 0.4.1.0 and 0.5 now at least.

sdroege commented:
Let me know if you would like anything else changed, thanks

stepcut commented:
I am not seeing the updated pull request -- only the original patch. Am I doing something wrong?

sdroege commented:
https://github.com/sdroege/acid-state/commit/0fc016b53e176e19bd07c0ba1bbc90016d58dacb this is the commit. The only change to the original pull request is that I dropped the change in the cabal file (originally I required >= 0.5, but after testing with 0.4 and seeing that it worked I dropped this now).

sdroege commented:
Thanks! :)
